File: 8.cpp
#include "aoc.hpp"

using namespace std;

const string inputname = "8.in";
using sol_t = int;
using instruction_t = std::pair<std::string,int>;
using program_t = std::vector<instruction_t>;

sol_t one(program_t & program)
{
    int acc = 0;
    int idx = 0;
    auto ni = program[idx];
    set<int> executed{};
    while (executed.find(idx)==executed.end())
    {
        if (program[idx].first == "acc")
        {
            acc += program[idx].second;
            executed.insert(idx);
            ++idx;
        }
        else if (program[idx].first == "nop")
        {
            executed.insert(idx);
            ++idx;
        }
        else if (program[idx].first == "jmp")
        {
            executed.insert(idx);
            idx+=program[idx].second;
        }
        else
        {
            std::cout << "INVALID INSTRUCTION AT LINE: " << idx << '\n';
        }
    }
    return acc;
}

std::pair<int,int> run_and_change(program_t program, int idx_c )
{
    if (program[idx_c].first == "nop")
    {
        program[idx_c].first = "jmp";
    }
    if (program[idx_c].first == "jmp")
    {
        program[idx_c].first = "nop";
    }
    int acc = 0;
    int idx = 0;
    int ni = program[idx].second;
    int li = idx;
    
    std::set<int> executed{};
    while ((executed.find(idx)==executed.end()) && (idx < program.size()) || (idx==(program.size()-1)))
    {
        if (program[idx].first == "acc")
        {
            acc += program[idx].second;
            executed.insert(idx);
            li = idx;
            ++idx;
        }
        else if (program[idx].first == "nop")
        {
            executed.insert(idx);
            li = idx;
            ++idx;
        }
        else if (program[idx].first == "jmp")
        {
            executed.insert(idx);
            li = idx;
            idx+=program[idx].second;
        }
        else
        {
            std::cout << "INVALID INSTRUCTION AT LINE: " << idx << '\n';
        }
    }
    return std::pair<int, int>{acc, li};
}

sol_t two(program_t & program)
{
    for (int i = 0; i < program.size()-1; ++i)
    {
        auto p = run_and_change(program, i);
        if (p.second==program.size()-1)
        {
            return p.first;
        }
    }
    return 0;
}

program_t program{};

int main(int argc, char **kwargs)
{
    ifstream infile(inputname);
    string s;
    while( getline( infile, s ) ) // fo r each line read from the file
    {
        auto pieces = split_string(s, ' ');
        instruction_t instruction{pieces[0],stoi(pieces[1])};
        program.push_back(instruction);
    }
    std::cout << "one = " << one(program) << '\n'; // 1217
    std::cout << "two = " << two(program) << '\n'; // 501

    return 0;
}

File: 8.py
from copy import deepcopy
import itertools
import re
from collections import namedtuple

ls = []
for s in open('8.in','r').readlines():
    x = s.rstrip('\n').split(' ')
    ls.append((x[0],int(x[1])))
    

def one(ls):
    acc = 0
    idx = 0
    ni = ls[idx]
    executed = set()
    while idx not in executed:
        if ls[idx][0] == 'acc':
            acc+= ls[idx][1]
            executed.add(idx)
            idx+=1
        elif ls[idx][0] == 'nop':
            executed.add(idx)
            idx+=1
        elif ls[idx][0] == 'jmp':
            executed.add(idx)
            idx+=ls[idx][1]
    return acc

def run_and_change(l, idx_c):
    ls = deepcopy(l)
    if ls[idx_c][0] == 'nop':
        ls[idx_c] = ('jmp', ls[idx_c][1])
    if ls[idx_c][0] == 'jmp':
        ls[idx_c] = ('nop', ls[idx_c][1])
    acc = 0
    idx = 0
    ni = ls[idx]
    li = idx
    executed = set()
    while (idx<len(ls) and idx not in executed) or idx==(len(ls)-1):
        if ls[idx][0] == 'acc':
            acc+= ls[idx][1]
            executed.add(idx)
            li = idx
            idx+=1
        elif ls[idx][0] == 'nop':
            executed.add(idx)
            li = idx
            idx+=1
        elif ls[idx][0] == 'jmp':
            executed.add(idx)
            li = idx
            idx+=ls[idx][1]
    return acc,li
    
def two(ls):
    for x in range(len(ls)-1):
        acc,idx = run_and_change(ls, x)
        if idx==(len(ls)-1): return acc
    return 0
    

if __name__ == "__main__":
    print(f"{one(ls) = }") # 1217
    print(f"{two(ls) = }") # 501
